### PR TITLE
Add ISmtpService extension methods

### PR DESCRIPTION
**Pull request by hishamco**

### CODE REVIEW

**Comment by hishamco** [comment]
@MikeAlhayek please review if you have time, just simple extension methods to simplify sending the emails

**Comment by MikeAlhayek** [comment]
@hishamco 
Can we please finish the other Email PR first and revisit this after?

**Comment by hishamco** [comment]
Seems the Azure Email will takes so long after the tons of feedback :)

File: src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs:23]: I don't think we should add all these extensions. I just don't think we'll used them. How about adding couple constructors to MailMessage instead?

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs:23]: I already saw such a pattern a long time ago in many frameworks & CMSs

> How about adding couple constructors to MailMessage instead?

Me as I used I will prefer a simple extension rather than instantiate a new `MailMessage`

Comment (Piedone) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs:23]: I think such shortcuts are only useful for the absolutely most frequent use cases. Otherwise, we'll duplicate the API.

This is especially important because unless you deliberately use named arguments, a method call will be less readable than an object initialization. See e.g. below; you have to put more effort into understanding which argument of `SendAsync()` does what, as opposed to obvious names:

image

Can we say something like "75% percent of `SendAsync()` calls only use `To`, `Subject`, and `Body` from `MailMessage`? Because then it's worthwhile adding an extension for that scenario.

**Review by Piedone — CHANGES_REQUESTED** [review]
This will need to wait for https://github.com/OrchardCMS/OrchardCore/pull/14715.

**Comment by hishamco** [comment]
I need to merge this if no one has any comment because I need to continue on PR #14715

File: src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs]: You didn't reply here.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs]: I wrote an extension for that

```csharp
    public static async Task<EmailResult> SendAsync(this IEmailService emailService, string to, string subject, string body, bool isHtmlBody = true)
    {
        var message = new MailMessage
        {
            To = to,
            Subject = subject,
            Body = body,
            IsHtmlBody = isHtmlBody
        };

        return await emailService.SendAsync(message);
    }
```

Comment (Piedone) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs]: I'm not convinced it's necessary or even beneficial though. Please address my https://github.com/OrchardCMS/OrchardCore/pull/14772#discussion_r1536077333 comment.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs]: > Can we say something like "75% percent of SendAsync() calls only use To, Subject, and Body from MailMessage?

I understand that the extension should have the above arguments, right? Or maybe I missundertood

Comment (Piedone) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs]: No, it's a question :).

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs]: I'm out of control because I'm working with many things :)

Yes we can say 80 - 90% uses the above arguments

File: src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs:6]:
```suggestion
/// Provides extension methods to <see cref="ISmtpService"/>.
```

File: src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs:17]:
```suggestion
    /// <param name="isHtmlBody">Whether the <paramref name="body"/> is in HTML format or not. Defaults to <c>true</c>.</param>
```

File: src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Email.Abstractions/Extensions/EmailServiceExtensions.cs:31]:
```suggestion
    public static Task<EmailResult> SendAsync(this IEmailService emailService, string to, string subject, string body, bool isHtmlBody = true)
    {
        var message = new MailMessage
        {
            To = to,
            Subject = subject,
            Body = body,
            IsHtmlBody = isHtmlBody
        };

        return emailService.SendAsync(message);
    }
```

File: src/OrchardCore/OrchardCore.Notifications.Core/Services/EmailNotificationProvider.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Notifications.Core/Services/EmailNotificationProvider.cs:35]:
```suggestion
        bool isHtmlBody;
        
```

**Review by Piedone — CHANGES_REQUESTED** [review]
I'm otherwise OK with this, @MikeAlhayek?

If we'll also merge https://github.com/OrchardCMS/OrchardCore/pull/14715 then this extension's signature will need to be changed too, but OK.

**Review by MikeAlhayek — APPROVED** [review]

**Comment by MikeAlhayek** [comment]
I am good with that one extension.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/EmailAuthenticatorController.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/EmailAuthenticatorController.cs:169]: I think the old version was nicer lol

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/EmailAuthenticatorController.cs:169]: I'm not sure why there's two async version, I might change this before I merge this PR

**Comment by hishamco** [comment]
@MikeAlhayek the last refactoring looks much better

**Review by Piedone — CHANGES_REQUESTED** [review]
I don't think the last commit brings any value but makes the code more complex. I'm OK with the code before that.

**Comment by hishamco** [comment]
> I don't think the last commit brings any value but makes the code more complex. I'm OK with the code before that.

I don't think so :) I'm not sure why we need two methods that do the exact thing except one for the subject and the other for the body

**Comment by Piedone** [comment]
The difference boils down to the fact that instead of having a `GetSubjectAsync` and `GetBodyAsync` separately, there's a tuple-returning `GetSubjectAndBodyAsync` that does it for both. This isn't better: The code does the same operations as before (i.e. no reduction) but instead of having a structured approeach of three specific methods, does all in one mixed together. The tuple is a code smell of this too.

**Comment by hishamco** [comment]
I will revert back to the last commit and then merge

**Review by Piedone — APPROVED** [review]

**Pull request merged**
